TauAnalysis/MCEmbeddingTools/plugins/DYToMuMuGenFilter.cc

```diff
@@ -0,0 +1,123 @@
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticleFwd.h"
+#include "DataFormats/Candidate/interface/LeafCandidate.h"
+#include "DataFormats/Candidate/interface/Candidate.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+
+#include "FWCore/Framework/interface/stream/EDFilter.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+
+
+class DYToMuMuGenFilter: public edm::stream::EDFilter<> {
+   public:
+      explicit DYToMuMuGenFilter(const edm::ParameterSet&);
+      ~DYToMuMuGenFilter();
+
+      static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+   private:
+      virtual void beginStream(edm::StreamID) override;
+      virtual bool filter(edm::Event&, const edm::EventSetup&) override;
+      virtual void endStream() override;
+      
+      edm::InputTag inputTag_;
+      edm::EDGetTokenT<reco::GenParticleCollection> genParticleCollection_;
+
+      //virtual void beginRun(edm::Run const&, edm::EventSetup const&) override;
+      //virtual void endRun(edm::Run const&, edm::EventSetup const&) override;
+      //virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+      //virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+
+      // ----------member data ---------------------------
+};
+
+
+
+
+DYToMuMuGenFilter::DYToMuMuGenFilter(const edm::ParameterSet& iConfig)
+{
+  inputTag_= iConfig.getParameter<edm::InputTag>("inputTag");
+  genParticleCollection_ = consumes<reco::GenParticleCollection>(inputTag_);
+}
+
+
+DYToMuMuGenFilter::~DYToMuMuGenFilter() {
+}
+
+
+bool DYToMuMuGenFilter::filter(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+
+
+    edm::Handle<reco::GenParticleCollection> gen_handle;
+    iEvent.getByToken(genParticleCollection_, gen_handle);
+    
+    for(unsigned int i = 0; i < gen_handle->size(); i++)
+    {
+		// Check if Z Boson decayed into two leptons
+		if (gen_handle->at(i).pdgId() == 23 && gen_handle->at(i).numberOfDaughters() == 2)
```

Comment:
"[i]" is more appropriate here, because by construction this loop is in the range of the collection size and the additional check made with .at is not necessary.
Also, get a const reference of the particle once and reuse it, instead of relying on the compiler to optimize out ```gen_handle->at(i) ```